File: static/cpp/2-exists.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <deque>
#include <iostream>
#include <random>
#include <vector>

using std::boolalpha;
using std::cout;
using std::default_random_engine;
using std::deque;
using std::endl;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

struct GraphVertex;
bool HasCycle(GraphVertex*, const GraphVertex*);

// @include
struct GraphVertex {
  enum Color { white, gray, black } color;
  vector<GraphVertex*> edges;
  // @exclude
  GraphVertex() : color(white) {}
  // @include
};

bool IsGraphMinimallyConnected(vector<GraphVertex>* G) {
  return G->empty() || !HasCycle(&G->front(), nullptr);
}

bool HasCycle(GraphVertex* cur, const GraphVertex* pre) {
  // Visiting a gray vertex means a cycle.
  if (cur->color == GraphVertex::gray) {
    return true;
  }

  cur->color = GraphVertex::gray;  // marks current vertex as a gray one.
  // Traverse the neighbor vertices.
  for (GraphVertex*& next : cur->edges) {
    if (next != pre && next->color != GraphVertex::black) {
      if (HasCycle(next, cur)) {
        return true;
      }
    }
  }
  cur->color = GraphVertex::black;  // marks current vertex as black.
  return false;
}
// @exclude

void HasCycleExclusion(GraphVertex* cur, GraphVertex* a, GraphVertex* b) {
  cur->color = GraphVertex::black;
  for (GraphVertex*& next : cur->edges) {
    if (next->color == GraphVertex::white &&
        ((cur != a && cur != b) || (next != a && next != b))) {
      HasCycleExclusion(next, a, b);
    }
  }
}

// O(n^2) check answer.
bool CheckAnswer(vector<GraphVertex>* G) {
  // marks all vertices as white.
  for (GraphVertex& n : *G) {
    n.color = GraphVertex::white;
  }

  for (GraphVertex& g : *G) {
    for (GraphVertex*& t : g.edges) {
      HasCycleExclusion(&g, &g, t);
      int count = 0;
      for (GraphVertex& n : *G) {
        if (n.color == GraphVertex::black) {
          ++count;
          n.color = GraphVertex::white;
        }
      }
      if (count == G->size()) {
        return false;
      }
    }
  }
  return true;
}

void TestUndirectedCycle() {
  vector<GraphVertex> G(3);
  G[0].edges.emplace_back(&G[2]);
  G[1].edges.emplace_back(&G[0]);
  G[2].edges.emplace_back(&G[1]);
  G[2].edges.emplace_back(&G[0]);
  G[0].edges.emplace_back(&G[1]);
  G[1].edges.emplace_back(&G[2]);
  bool result = IsGraphMinimallyConnected(&G);
  cout << boolalpha << result << endl;
  assert(CheckAnswer(&G) == result);
}

void TestUndirectedStarTree() {
  vector<GraphVertex> G(4);
  G[0].edges.emplace_back(&G[1]);
  G[1].edges.emplace_back(&G[0]);
  G[0].edges.emplace_back(&G[2]);
  G[2].edges.emplace_back(&G[0]);
  G[0].edges.emplace_back(&G[3]);
  G[3].edges.emplace_back(&G[0]);
  bool result = IsGraphMinimallyConnected(&G);
  cout << boolalpha << result << endl;
  assert(CheckAnswer(&G) == result);
  assert(true == result);
}

void TestUndirectedLineTree() {
  vector<GraphVertex> G(4);
  G[0].edges.emplace_back(&G[1]);
  G[1].edges.emplace_back(&G[0]);
  G[1].edges.emplace_back(&G[2]);
  G[2].edges.emplace_back(&G[1]);
  G[2].edges.emplace_back(&G[3]);
  G[3].edges.emplace_back(&G[2]);
  bool result = IsGraphMinimallyConnected(&G);
  cout << boolalpha << result << endl;
  assert(CheckAnswer(&G) == result);
  assert(true == result);
  G[1].edges.emplace_back(&G[3]);
  G[3].edges.emplace_back(&G[1]);
  result = IsGraphMinimallyConnected(&G);
  assert(false == result);
}

void TestUndirectedBinaryTree() {
  vector<GraphVertex> G(7);
  G[0].edges.emplace_back(&G[1]);
  G[1].edges.emplace_back(&G[0]);
  G[0].edges.emplace_back(&G[2]);
  G[2].edges.emplace_back(&G[0]);
  G[1].edges.emplace_back(&G[3]);
  G[3].edges.emplace_back(&G[1]);
  G[1].edges.emplace_back(&G[4]);
  G[4].edges.emplace_back(&G[1]);
  G[2].edges.emplace_back(&G[5]);
  G[5].edges.emplace_back(&G[2]);
  G[2].edges.emplace_back(&G[6]);
  G[6].edges.emplace_back(&G[2]);
  bool result = IsGraphMinimallyConnected(&G);
  cout << boolalpha << result << endl;
  assert(CheckAnswer(&G) == result);
  assert(true == result);
  G[4].edges.emplace_back(&G[6]);
  G[6].edges.emplace_back(&G[4]);
  result = IsGraphMinimallyConnected(&G);
  assert(false == result);
}

void TestUndirectedTwoSeparateCycles() {
  vector<GraphVertex> G(6);
  G[0].edges.emplace_back(&G[1]);
  G[1].edges.emplace_back(&G[0]);
  G[1].edges.emplace_back(&G[2]);
  G[2].edges.emplace_back(&G[1]);
  G[2].edges.emplace_back(&G[0]);
  G[0].edges.emplace_back(&G[2]);
  G[3].edges.emplace_back(&G[4]);
  G[4].edges.emplace_back(&G[3]);
  G[4].edges.emplace_back(&G[5]);
  G[5].edges.emplace_back(&G[4]);
  G[5].edges.emplace_back(&G[3]);
  G[3].edges.emplace_back(&G[5]);
  bool result = IsGraphMinimallyConnected(&G);
  // since we just check for a cycle, result is false.
  // this is ok, because we assume input is connected.
  assert(false == result);
}

int main(int argc, char* argv[]) {
  TestUndirectedCycle();
  TestUndirectedStarTree();
  TestUndirectedLineTree();
  TestUndirectedBinaryTree();
  TestUndirectedTwoSeparateCycles();
  default_random_engine gen((random_device())());
  for (int times = 0; times < 100; ++times) {
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(2, 2000);
      n = dis(gen);
    }
    vector<GraphVertex> G(n);
    uniform_int_distribution<int> dis(1, n * (n - 1) / 2);
    int m = dis(gen);
    vector<deque<bool>> is_edge_exist(n, deque<bool>(n, false));
    // Make the graph become connected.
    for (int i = 1; i < n; ++i) {
      G[i - 1].edges.emplace_back(&G[i]);
      G[i].edges.emplace_back(&G[i - 1]);
      is_edge_exist[i - 1][i] = is_edge_exist[i][i - 1] = true;
    }
    // Generate edges randomly.
    m -= (n - 1);
    while (m-- > 0) {
      int a, b;
      do {
        uniform_int_distribution<int> dis(0, n - 1);
        a = dis(gen), b = dis(gen);
      } while (a == b || is_edge_exist[a][b] == true);
      is_edge_exist[a][b] = is_edge_exist[b][a] = true;
      G[a].edges.emplace_back(&G[b]);
      G[b].edges.emplace_back(&G[a]);
    }

    bool result = IsGraphMinimallyConnected(&G);
    cout << boolalpha << result << endl;
    assert(CheckAnswer(&G) == result);
  }
  return 0;
}

File: static/python/2_exists.py
import sys
import random


# @include
class GraphVertex:

    white, gray, black = range(3)

    def __init__(self):
        self.color = GraphVertex.white
        self.edges = []
# @exclude

    def __repr__(self):
        return '(%d)%d(%s)' % (self.color, id(self), ','.join(
            str(id(x)) for x in self.edges))


# @include


def is_deadlocked(G):
    def has_cycle(cur):
        # Visiting a gray vertex means a cycle.
        if cur.color == GraphVertex.gray:
            return True

        cur.color = GraphVertex.gray  # Marks current vertex as a gray one.
        # Traverse the neighbor vertices.
        if any(next.color != GraphVertex.black and has_cycle(next)
               for next in cur.edges):
            return True
        cur.color = GraphVertex.black  # Marks current vertex as black.
        return False

    return any(vertex.color == GraphVertex.white and has_cycle(vertex)
               for vertex in G)
# @exclude


def has_cycle_exclusion(cur):
    if cur.color == GraphVertex.black:
        return True
    cur.color = GraphVertex.black
    return any(has_cycle_exclusion(next) for next in cur.edges)


# O(n^2) check answer.
def check_answer(G):
    # marks all vertices as white.
    for n in G:
        n.color = GraphVertex.white

    for g in G:
        if has_cycle_exclusion(g):
            return True
        # Reset color to white.
        for n in G:
            n.color = GraphVertex.white
    return False


def test_two_nodes_cycle():
    G = [GraphVertex() for _ in range(2)]
    G[0].edges.append(G[1])
    G[1].edges.append(G[0])
    result = is_deadlocked(G)
    print(result)
    assert check_answer(G) == result
    assert result


def test_directed_cycle():
    G = [GraphVertex() for _ in range(3)]
    G[0].edges.append(G[1])
    G[1].edges.append(G[2])
    G[2].edges.append(G[0])
    result = is_deadlocked(G)
    print(result)
    assert check_answer(G) == result
    assert result


def test_directed_star_tree():
    G = [GraphVertex() for _ in range(4)]
    G[0].edges.append(G[1])
    G[0].edges.append(G[2])
    G[0].edges.append(G[3])
    result = is_deadlocked(G)
    print(result)
    assert check_answer(G) == result
    assert not result


def test_directed_line_tree():
    G = [GraphVertex() for _ in range(4)]
    G[0].edges.append(G[1])
    G[1].edges.append(G[2])
    G[2].edges.append(G[3])
    result = is_deadlocked(G)
    print(result)
    assert check_answer(G) == result
    assert not result
    G[3].edges.append(G[1])
    result = is_deadlocked(G)
    assert result


def test_directed_binary_tree():
    G = [GraphVertex() for _ in range(7)]
    G[0].edges.append(G[1])
    G[0].edges.append(G[2])
    G[1].edges.append(G[3])
    G[1].edges.append(G[4])
    G[2].edges.append(G[5])
    G[2].edges.append(G[6])
    result = is_deadlocked(G)
    print(result)
    assert check_answer(G) == result
    assert not result
    G[4].edges.append(G[6])
    G[6].edges.append(G[1])
    result = is_deadlocked(G)
    assert result


def test_directed_two_separate_cycles():
    G = [GraphVertex() for _ in range(6)]
    G[0].edges.append(G[1])
    G[1].edges.append(G[2])
    G[2].edges.append(G[0])
    G[3].edges.append(G[4])
    G[4].edges.append(G[5])
    G[5].edges.append(G[3])
    result = is_deadlocked(G)
    assert result


def main():
    test_two_nodes_cycle()
    test_directed_cycle()
    test_directed_star_tree()
    test_directed_line_tree()
    test_directed_binary_tree()
    test_directed_two_separate_cycles()
    for _ in range(100):
        n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(2, 2000)
        G = [GraphVertex() for _ in range(n)]
        m = random.randint(1, n * (n - 1) // 2)
        is_edge_exist = [[False] * n for _ in range(n)]
        # Make the graph become connected.
        for i in range(1, n):
            G[i - 1].edges.append(G[i])
            G[i].edges.append(G[i - 1])
            is_edge_exist[i - 1][i] = is_edge_exist[i][i - 1] = True

        # Generate edges randomly.
        m -= (n - 1)
        while m > 0:
            m -= 1
            while True:
                a = random.randrange(n)
                b = random.randrange(n)
                if a != b and is_edge_exist[a][b] == False:
                    break
            is_edge_exist[a][b] = is_edge_exist[b][a] = True
            G[a].edges.append(G[b])
            G[b].edges.append(G[a])

        result = is_deadlocked(G)
        print(result)
        assert check_answer(G) == result


if __name__ == '__main__':
    main()
